In [ ]:
import pandas as pd
import numpy as np
from matplotlib import pyplot as plt
from sklearn.preprocessing import OneHotEncoder

!pip install mlxtend==0.23.1

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


# Association Rule for Store Dataset

In this case study, we will explore how association rule can be used to analyze the items that are usualy purcased together.

you can refer to this article to find out about apriori and association rule:
https://rasbt.github.io/mlxtend/user_guide/frequent_patterns/apriori/
https://rasbt.github.io/mlxtend/user_guide/frequent_patterns/association_rules/

## Load Data

We will use the dataset of the transaction in a certain store. You can get the dataset here:
https://gist.githubusercontent.com/Harsh-Git-Hub/2979ec48043928ad9033d8469928e751/raw/72de943e040b8bd0d087624b154d41b2ba9d9b60/retail_dataset.csv

In [ ]:
url = "https://gist.githubusercontent.com/Harsh-Git-Hub/2979ec48043928ad9033d8469928e751/raw/72de943e040b8bd0d087624b154d41b2ba9d9b60/retail_dataset.csv"
df = pd.read_csv(url)
df.head()

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


0       1     2       3       4       5       6
0   Bread    Wine  Eggs    Meat  Cheese  Pencil  Diaper
1   Bread  Cheese  Meat  Diaper    Wine    Milk  Pencil
2  Cheese    Meat  Eggs    Milk    Wine     NaN     NaN
3  Cheese    Meat  Eggs    Milk    Wine     NaN     NaN
4    Meat  Pencil  Wine     NaN     NaN     NaN     NaN

# Get the set of product that has been purchased


In [ ]:
print(set(np.ravel(df.values)))`

{nan, 'Diaper', 'Meat', 'Eggs', 'Cheese', 'Bagel', 'Bread', 'Wine', 'Milk', 'Pencil'}


/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


## Preprocess Data

In this step, we will transform our dataset so that we will have a one hot encoding based on the purchased products.

In [ ]:
unique = pd.unique(df.values.ravel())
encoding = {product: 1 if product not in ['Milk', 'Bagel', np.nan] else 0 for product in unique}

print(encoding)

{'Bread': 1, 'Wine': 1, 'Eggs': 1, 'Meat': 1, 'Cheese': 1, 'Pencil': 1, 'Diaper': 1, 'Milk': 0, nan: 0, 'Bagel': 0}


/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [ ]:
unique_products = pd.unique(df.values.ravel())

encoded_df = pd.DataFrame(0, index=df.index, columns=unique_products)

for index, row in df.iterrows():
    for product in row:
        if product in encoded_df.columns:
            encoded_df.at[index, product] = 1

encoded_df.head()

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


Bread  Wine  Eggs  Meat  Cheese  Pencil  Diaper  Milk  NaN  Bagel
0      1     1     1     1       1       1       1     0    0      0
1      1     1     0     1       1       1       1     1    0      0
2      0     1     1     1       1       0       0     1    1      0
3      0     1     1     1       1       0       0     1    1      0
4      0     1     0     1       0       1       0     0    1      0

In [ ]:
encoded_df = encoded_df.drop(columns=[np.nan])

encoded_df.head()

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


Bread  Wine  Eggs  Meat  Cheese  Pencil  Diaper  Milk  Bagel
0      1     1     1     1       1       1       1     0      0
1      1     1     0     1       1       1       1     1      0
2      0     1     1     1       1       0       0     1      0
3      0     1     1     1       1       0       0     1      0
4      0     1     0     1       0       1       0     0      0

Since, the encoded dataframe consist of the empty column. We will drop the NaN column or select all columns other than the first column.

## Apriori Algorithm

We will use appriori algorithm to determine the frequently purchased products.
For this case study, we will min_support=0.2

In [ ]:
from mlxtend.frequent_patterns import apriori, association_rules

frequent_itemsets = apriori(encoded_df, min_support=0.2, use_colnames=True)
frequent_itemsets


/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)
/usr/local/lib/python3.10/dist-packages/mlxtend/frequent_patterns/fpcommon.py:109: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(


support              itemsets
0   0.504762               (Bread)
1   0.438095                (Wine)
2   0.438095                (Eggs)
3   0.476190                (Meat)
4   0.501587              (Cheese)
5   0.361905              (Pencil)
6   0.406349              (Diaper)
7   0.501587                (Milk)
8   0.425397               (Bagel)
9   0.244444         (Bread, Wine)
10  0.206349         (Bread, Meat)
11  0.238095       (Cheese, Bread)
12  0.200000       (Bread, Pencil)
13  0.231746       (Diaper, Bread)
14  0.279365         (Milk, Bread)
15  0.279365        (Bagel, Bread)
16  0.241270          (Eggs, Wine)
17  0.250794          (Meat, Wine)
18  0.269841        (Cheese, Wine)
19  0.200000        (Wine, Pencil)
20  0.234921        (Diaper, Wine)
21  0.219048          (Milk, Wine)
22  0.266667          (Eggs, Meat)
23  0.298413        (Eggs, Cheese)
24  0.244444          (Eggs, Milk)
25  0.323810        (Cheese, Meat)
26  0.244444          (Milk, Meat)
27  0.200000      (Cheese, Pencil)
28  0.200000      (Cheese, Diaper)
29  0.304762        (Cheese, Milk)
30  0.225397         (Bagel, Milk)
31  0.215873  (Eggs, Cheese, Meat)
32  0.203175  (Cheese, Milk, Meat)

The we will generate association rule of the frequent itemset based on confidence level with the threshold=0.6

In [ ]:
association_rules(frequent_itemsets, metric="confidence", min_threshold=0.6)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


antecedents consequents  antecedent support  consequent support  \
0          (Bagel)     (Bread)            0.425397            0.504762   
1           (Wine)    (Cheese)            0.438095            0.501587   
2           (Eggs)      (Meat)            0.438095            0.476190   
3           (Eggs)    (Cheese)            0.438095            0.501587   
4         (Cheese)      (Meat)            0.501587            0.476190   
5           (Meat)    (Cheese)            0.476190            0.501587   
6         (Cheese)      (Milk)            0.501587            0.501587   
7           (Milk)    (Cheese)            0.501587            0.501587   
8   (Eggs, Cheese)      (Meat)            0.298413            0.476190   
9     (Eggs, Meat)    (Cheese)            0.266667            0.501587   
10  (Cheese, Meat)      (Eggs)            0.323810            0.438095   
11  (Cheese, Milk)      (Meat)            0.304762            0.476190   
12  (Cheese, Meat)      (Milk)            0.323810            0.501587   
13    (Milk, Meat)    (Cheese)            0.244444            0.501587   

     support  confidence      lift  leverage  conviction  zhangs_metric  
0   0.279365    0.656716  1.301042  0.064641    1.442650       0.402687  
1   0.269841    0.615942  1.227986  0.050098    1.297754       0.330409  
2   0.266667    0.608696  1.278261  0.058050    1.338624       0.387409  
3   0.298413    0.681159  1.358008  0.078670    1.563203       0.469167  
4   0.323810    0.645570  1.355696  0.084958    1.477891       0.526414  
5   0.323810    0.680000  1.355696  0.084958    1.557540       0.500891  
6   0.304762    0.607595  1.211344  0.053172    1.270148       0.350053  
7   0.304762    0.607595  1.211344  0.053172    1.270148       0.350053  
8   0.215873    0.723404  1.519149  0.073772    1.893773       0.487091  
9   0.215873    0.809524  1.613924  0.082116    2.616667       0.518717  
10  0.215873    0.666667  1.521739  0.074014    1.685714       0.507042  
11  0.203175    0.666667  1.400000  0.058050    1.571429       0.410959  
12  0.203175    0.627451  1.250931  0.040756    1.337845       0.296655  
13  0.203175    0.831169  1.657077  0.080564    2.952137       0.524816

Provide explanation about __antecedent support__, __consequent support__, __support__, __confidence__, __lift__, __leverage__, __conviction__, __conviction__ and the interpretation from the case above (please use text section)


Antecedent support mengukur seberapa sering item atau kombinasi item di sisi kiri aturan muncul dalam dataset, misalnya, "Bagel" muncul sekitar 42.5% dalam dataset. Consequent support mengukur seberapa sering item di sisi kanan aturan muncul, seperti "Bread" yang muncul 50.5% dalam transaksi. Support mengukur seberapa sering pasangan antecedent dan consequent muncul bersamaan, contohnya, "Bagel → Bread" muncul 27.9% dari transaksi.

Confidence menunjukkan seberapa besar kemungkinan consequent muncul jika antecedent ada, contohnya, jika "Bagel" ada, maka ada 65.7% kemungkinan "Bread" juga ada. Lift membandingkan probabilitas munculnya kedua item bersama-sama dengan probabilitas acak mereka muncul, dan nilai lebih besar dari 1 (seperti 1.3) menunjukkan keduanya lebih sering muncul bersama daripada yang diharapkan secara acak. Leverage mengukur peningkatan yang terjadi pada support bersama dibandingkan dengan ekspektasi jika keduanya muncul secara independen, yang pada contoh ini menunjukkan peningkatan 6.5%.

Conviction mengukur seberapa kuat hubungan antara antecedent dan consequent, dan nilai lebih besar dari 1 (seperti 1.44) menunjukkan hubungan yang kuat. Zhang’s metric adalah metrik lain yang mengevaluasi seberapa relevan aturan asosiasi tersebut, dengan nilai 0.40 menunjukkan aturan tersebut memiliki kekuatan yang cukup baik, meskipun tidak terlalu kuat.

Secara keseluruhan, aturan asosiasi yang ditemukan menunjukkan hubungan yang signifikan antara produk-produk tertentu, seperti "Bagel" dan "Bread", yang muncul lebih sering bersama-sama daripada yang diharapkan secara acak.